java/ql/lib/semmle/code/java/dataflow/internal/DataFlowUtil.qll

```diff
@@ -190,14 +190,21 @@ predicate simpleAstFlowStep(Expr e1, Expr e2) {
   e2 = any(NotNullExpr nne | e1 = nne.getExpr())
   or
   e2.(WhenExpr).getBranch(_).getAResult() = e1
+  or
+  exists(SwitchExpr se | e1 = se.getExpr() and e2 = se.getACase().(PatternCase).getPattern())
+  or
+  exists(SwitchStmt ss | e1 = ss.getExpr() and e2 = ss.getACase().(PatternCase).getPattern())
+  or
+  exists(InstanceOfExpr ioe | e1 = ioe.getExpr() and e2 = ioe.getPattern())
```

Comment:
I think we generally want the flow for simple patterns to be included in `VariableAssign`. That means that the `e1` expression here steps directly through the ssa definition to the first use via `simpleLocaLFlowStep0` below, so we should avoid creating parallel paths in those cases and thus exclude simple patterns in these three disjuncts.
Adding support in `VariableAssign.getSource` will also cover the currently missing case in `closureFlowStep` in `DataFlowPrivate.qll`.

Reply:
There's another missing flow step in `localMustFlowStep`, which will also be handled once `VariableAssign.getSource` covers simple patterns.

Reply:
I've amended these disjuncts to only care about record patterns, where we need an intermediate node to terminate the read step (or messily use ssa-step in defining the read step).

For example, simple pattern `x instanceof T v ... use(v)` now looks like `x -flow-> v_use`, but `x instanceof R1(R2(t v)) ... use(v)` looks like `x -flow-> R1 -read-> R2 -read-> v -flow-> v_use`.